src/Storages/StoragePostgreSQL.cpp

```diff
@@ -0,0 +1,322 @@
+        IdentifierQuotingStyle::DoubleQuotes, "", remote_table_name, context_);
+
+    Block sample_block;
+    for (const String & column_name : column_names_)
+    {
+        auto column_data = metadata_snapshot->getColumns().getPhysical(column_name);
+        WhichDataType which(column_data.type);
+        if (which.isEnum())
+            column_data.type = std::make_shared<DataTypeString>();
+        sample_block.insert({ column_data.type, column_data.name });
+    }
+
+    return Pipe(std::make_shared<SourceFromInputStream>(
+            std::make_shared<PostgreSQLBlockInputStream>(connection->conn(), query, sample_block, max_block_size_)));
+}
+
+
+class PostgreSQLBlockOutputStream : public IBlockOutputStream
+{
+public:
+    explicit PostgreSQLBlockOutputStream(
+        const StorageMetadataPtr & metadata_snapshot_,
+        ConnectionPtr connection_,
+        const std::string & remote_table_name_)
+        : metadata_snapshot(metadata_snapshot_)
+        , connection(connection_)
+        , remote_table_name(remote_table_name_)
+    {
+    }
+
+    Block getHeader() const override { return metadata_snapshot->getSampleBlock(); }
+
+
+    void writePrefix() override
+    {
+        work = std::make_unique<pqxx::work>(*connection);
+    }
+
+
+    void write(const Block & block) override
+    {
+        if (!work)
+            return;
+
+        const auto columns = block.getColumns();
+        const size_t num_rows = block.rows(), num_cols = block.columns();
+        const auto data_types = block.getDataTypes();
+
+        if (!stream_inserter)
+            stream_inserter = std::make_unique<pqxx::stream_to>(*work, remote_table_name, block.getNames());
+
+        /// std::optional lets libpqxx to know if value is NULL
+        std::vector<std::optional<std::string>> row(num_cols);
+
+        for (const auto i : ext::range(0, num_rows))
+        {
+            for (const auto j : ext::range(0, num_cols))
+            {
+                if (columns[j]->isNullAt(i))
+                {
+                    row[j] = std::nullopt;
+                }
+                else
+                {
+                    WriteBufferFromOwnString ostr;
+
+                    if (isArray(data_types[j]))
+                    {
+                        parseArray((*columns[j])[i], data_types[j], ostr);
+                    }
+                    else
+                    {
+                        data_types[j]->serializeAsText(*columns[j], i, ostr, FormatSettings{});
+                    }
+
+                    row[j] = ostr.str();
+                }
+            }
+
+            stream_inserter->write_values(row);
+        }
+    }
+
+
+    void writeSuffix() override
+    {
+        if (stream_inserter)
+        {
+            stream_inserter->complete();
+            work->commit();
+        }
+    }
+
+
+    /// Cannot just use serializeAsText for array data type even though it converts perfectly
```

Comment:
BTW Postgres support `array[1, 2, 3]` syntax:
http://sqlfiddle.com/#!17/0a28f/359

Reply:
But it fails with copy command:

E           helpers.client.QueryRuntimeException: Client failed! Return code: 233, stderr: Received exception from server (version 21.1.1):
E           Code: 1001. DB::Exception: Received from 172.21.0.3:9000. DB::Exception: pqxx::data_exception: ERROR:  malformed array literal: "array['2000-05-12','2000-05-12']"
E           DETAIL:  Array value must start with "{" or dimension information.
E           CONTEXT:  COPY test_array_dimensions, line 1, column a: "array['2000-05-12','2000-05-12']"

E           helpers.client.QueryRuntimeException: Client failed! Return code: 233, stderr: Received exception from server (version 21.1.1):
E           Code: 1001. DB::Exception: Received from 172.21.0.3:9000. DB::Exception: pqxx::data_exception: ERROR:  malformed array literal: "array[[[1,1],[1,1]],[[3,3],[3,3]],[[4,4],[5,5]]]"
E           DETAIL:  Array value must start with "{" or dimension information.
E           CONTEXT:  COPY array_columns, line 1, column b: "array[[[1,1],[1,1]],[[3,3],[3,3]],[[4,4],[5,5]]]"